lv_draw/lv_draw_label.c

```diff
@@ -54,148 +52,214 @@ static uint8_t hex_char_to_num(char hex);
-                if(cmd_state == CMD_STATE_PAR) {
-                    if(letter == ' ') {
-                        /*Get the parameter*/
-                        if(i - par_start == LABEL_RECOLOR_PAR_LENGTH + 1) {
-                            char buf[LABEL_RECOLOR_PAR_LENGTH + 1];
-                            memcpy(buf, &txt[par_start], LABEL_RECOLOR_PAR_LENGTH);
-                            buf[LABEL_RECOLOR_PAR_LENGTH] = '\0';
-                            int r, g, b;
-                            r = (hex_char_to_num(buf[0]) << 4) + hex_char_to_num(buf[1]);
-                            g = (hex_char_to_num(buf[2]) << 4) + hex_char_to_num(buf[3]);
-                            b = (hex_char_to_num(buf[4]) << 4) + hex_char_to_num(buf[5]);
-                            recolor = LV_COLOR_MAKE(r, g, b);
-                        } else {
-                            recolor.full = style->text.color.full;
-                        }
-                        cmd_state = CMD_STATE_IN; /*After the parameter the text is in the command*/
-                    }
-                    continue;
-                }
-            }
-
-            lv_color_t color = style->text.color;
-
-            if(cmd_state == CMD_STATE_IN) color = recolor;
-
-            letter_fp(&pos, mask, font, letter, color, opa);
-            letter_w = lv_font_get_width(font, letter);
-
-            pos.x += letter_w + style->text.letter_space;
-        }
-        /*Go to next line*/
-        line_start = line_end;
-        line_end += lv_txt_get_next_line(&txt[line_start], font, style->text.letter_space, w, flag);
-
-        pos.x = coords->x1;
-        /*Align to middle*/
-        if(flag & LV_TXT_FLAG_CENTER) {
-            line_width = lv_txt_get_width(&txt[line_start], line_end - line_start,
-                                          font, style->text.letter_space, flag);
-
-            pos.x += (lv_area_get_width(coords) - line_width) / 2;
-
-        }
-        /*Align to the right*/
-        else if(flag & LV_TXT_FLAG_RIGHT) {
-            line_width = lv_txt_get_width(&txt[line_start], line_end - line_start,
-                                          font, style->text.letter_space, flag);
-            pos.x += lv_area_get_width(coords) - line_width;
-        }
-
-        /*Go the next line position*/
-        pos.y += lv_font_get_height(font);
-        pos.y += style->text.line_space;
-    }
+	/*Write out all lines*/
+	while (txt[line_start] != '\0') {
+		if (offset != NULL) {
+			pos.x += x_ofs;
+		}
+		/*Write all letter of a line*/
+		cmd_state = CMD_STATE_WAIT;
+		i = line_start;
+		uint32_t letter;
+		// my code
+		uint32_t previous_letter;
+		uint32_t current_letter;
+		reversed_index = 0;
+		uint8_t noWidth = 0;
+		// end my code
+		while (i < line_end) {
+			// my code
+			//if it's necessary make the txt buffer reversed
+			if (!reversed_index) {
+				reversed_buffer = lv_get_reversed_buffer((uint8_t*) txt, i,
```

Comment:
`lv_get_reversed_buffer` will be called frequently, also when there is no Persian letter in the text. I think there should be a way to opt out and avoid this extra processing when developing an application exclusively for LTR languages. This can easily be done by a global `#define` that enables/disabled RTL processing.

Reply:
If I wanted to do this there must be an attribute that should be set for writing RTL and Right now I am working on a product that should support English and Farsi at the same time.
If you look at the function's body you would see this function only allocate buffer if it is necessary and if it is not necessary the pointer that was passed to  function will be returned without allocating new pointer. So I think this is not making considerable overhead for Processor